src/gflownet/algo/trajectory_balance.py

```diff
@@ -254,10 +267,45 @@ def compute_batch_losses(self, model: TrajectoryBalanceModel, batch: gd.Batch, n
             'invalid_trajectories': invalid_mask.sum() / batch.num_online if batch.num_online > 0 else 0,
             'invalid_logprob': (invalid_mask * traj_log_prob).sum() / (invalid_mask.sum() + 1e-4),
             'invalid_losses': (invalid_mask * traj_losses).sum() / (invalid_mask.sum() + 1e-4),
-            'logZ': Z.mean(),
+            'logZ': log_Z.mean(),
             'loss': loss.item(),
         }
 
         if not torch.isfinite(traj_losses).all():
             raise ValueError('loss is not finite')
         return loss, info
+
+    def _init_subtb(self, dev):
+        # Precompute all possible subtrajectory indices that we will use for computing the loss
+        ar = torch.arange(self._subtb_max_len, device=dev)
+        tidx = [torch.tril_indices(i, i, device=dev)[1] for i in range(self._subtb_max_len)]
+        self._precomp = [
+            (torch.cat([i + tidx[T - i]
+                        for i in range(T)]),
+             torch.cat([ar[:T - i].repeat_interleave(ar[:T - i] + 1) + ar[T - i + 1:T + 1].sum()
+                        for i in range(T)]))
+            for T in range(1, self._subtb_max_len)
+        ]
```

Comment:
I think it would be helpful to extend the docstring and say in words what this code is doing.